snapcraft/internal/pluginhandler/_runner.py

```diff
@@ -0,0 +1,193 @@
+        :param str builddir: The build directory for this part.
+        :param dict builtin_functions: Dict of builtin function names to
+                                       actual callables.
+        """
+        self._builddir = builddir
+        self._builtin_functions = builtin_functions
+
+        self._prepare_scriptlet = part_properties.get('prepare')
+        self._build_scriptlet = part_properties.get('build')
+        self._install_scriptlet = part_properties.get('install')
+
+    def prepare(self) -> None:
+        """Run prepare scriptlet."""
+        if self._prepare_scriptlet:
+            self._run_scriptlet(
+                'prepare', self._prepare_scriptlet, self._builddir)
+
+    def build(self) -> None:
+        """Run build scriptlet."""
+        if self._build_scriptlet:
+            self._run_scriptlet(
+                'build', self._build_scriptlet, self._builddir)
+
+    def install(self) -> None:
+        """Run install scriptlet."""
+        if self._install_scriptlet:
+            self._run_scriptlet(
+                'install', self._install_scriptlet, self._builddir)
+
+    def _run_scriptlet(self, scriptlet_name: str, scriptlet: str,
+                       workdir: str) -> None:
+        with tempfile.TemporaryDirectory() as tempdir:
+            call_fifo = _NonBlockingFifo(
+                path=os.path.join(tempdir, 'function_call'),
+                permissions=os.O_RDONLY)
+            feedback_fifo = _NonBlockingFifo(
+                path=os.path.join(tempdir, 'call_feedback'),
+                permissions=os.O_WRONLY)
+
+            script = textwrap.dedent("""\
+                {env}
+                {functions}
+                {scriptlet}
+            """.format(
+                env=common.assemble_env(), scriptlet=scriptlet,
+                functions=functions(
+                    call_fifo=call_fifo.path,
+                    feedback_fifo=feedback_fifo.path)))
+
+            process = subprocess.Popen(
+                ['/bin/sh', '-e', '-c', script], cwd=self._builddir)
+
+            status = None
+            try:
+                while status is None:
+                    function_call = call_fifo.read()
+                    if function_call:
+                        self._handle_builtin_function(
+                            scriptlet_name, function_call.strip())
+                        feedback_fifo.write('success\n')
+                    status = process.poll()
+            finally:
+                call_fifo.close()
+                feedback_fifo.close()
+
+            if status:
+                raise errors.ScriptletRunError(
+                    scriptlet_name=scriptlet_name, code=status)
+
+    def _handle_builtin_function(self, scriptlet_name, function_call):
+        function_call_components = shlex.split(function_call)
+        if len(function_call_components) == 0:
+            # This means a snapcraft developer screwed up. Should never be
```

Comment:
let's try and avoid the word "screwed", it feels a bit confrontational :-)

Reply:
Of course, good call. Fixed.